### PR TITLE
Improve TLS1.2 client authentication to use TSIP

### PR DESCRIPTION
**Pull request by TakayukiMatsuo**
# Description

This PR add support client authentication in TLS1.2 with TSIP acceleration. Both client certificates including RSA-2048/ ECC-P256 key can be handled.

# Testing

Tested by running example clients on following Renesas boards:
- RX72N Envision kit
- RSK-RX65N-2MB
- GR-ROSE

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [☑] updated appropriate READMEs
 - [☑] Updated manual and documentation

### CODE REVIEW

**Review by cconlon — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/port/Renesas/renesas_common.c [hunk 1]
```diff
@@ -74,6 +74,29 @@ WOLFSSL_LOCAL int Renesas_cmn_RsaSignCb(WOLFSSL* ssl,
     WOLFSSL_LEAVE("Renesas_cmn_RsaSignCb", ret);
     return ret;
 }
+/* This function is a callback passed to wolfSSL_CTX_SetRsaSignCheckCb.
+ * And tries to verify the signature passed to it by decrypting with a public
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/port/Renesas/renesas_common.c]: Rephrased it.

File: wolfcrypt/src/port/Renesas/renesas_tsip_util.c [hunk 2]
```diff
@@ -3336,6 +3494,379 @@ int wc_tsip_tls_RootCertVerify(
...
+                    out = info->pk.eccsign.out;
+                    sig = sigData.pdata;
+
+                    rSz =  sSz = R_TSIP_ECDSA_DATA_BYTE_SIZE / 2;
+                    rSz += (sig[0] & 0x80)?1:0;
+                    sSz += (sig[sSz] & 0x80)?1:0;
+                    sz  =  (ASN_TAG_SZ + 1) * 2 + rSz + sSz;
+
+                    /* encode ASN sequence */
+                    out[idx++] = ASN_SEQUENCE | ASN_CONSTRUCTED;
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/port/Renesas/renesas_tsip_util.c:3777]: The buffer size is checked immediately after the case statement that precedes this statement.

File: wolfssl/wolfcrypt/port/Renesas/renesas-tsip-crypt.h [hunk 3]
```diff
@@ -79,6 +79,12 @@ enum {
     TSIP_TEMP_WORK_SIZE        = 128,
 };
 
+typedef enum {
+    TSIP_KEY_TYPE_RSA2048      = 0,
+    TSIP_KEY_TYPE_RSA4096      = 1,
+    TSIP_KEY_TYPE_ECDSAP256    = 2,
+}wolfssl_TSIP_KEY_TYPE;
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfssl/wolfcrypt/port/Renesas/renesas-tsip-crypt.h]: Fixed it as suggested.

File: wolfcrypt/src/port/Renesas/renesas_common.c [hunk 4]
```diff
@@ -181,6 +204,23 @@ static int Renesas_cmn_CryptoDevCb(int devIdArg, wc_CryptoInfo* info, void* ctx)
     #endif /* HAVE_AES_CBC */
     #endif /* !NO_AES || !NO_DES3 */
     }
+    /* be called for signing
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/port/Renesas/renesas_common.c]: Changed as suggested.

File: wolfcrypt/src/port/Renesas/renesas_tsip_util.c [hunk 5]
```diff
@@ -1688,8 +1767,83 @@ static int tsipImportPrivateKey(TsipUserCtx* tuc, const byte* encPrivKey,
...
+                if (err == TSIP_SUCCESS) {
+                    tuc->ClientRsa2048PubKey_set = 1;
+                }
+                else {
+                    ret = WC_HW_E;
+                }
+                break;
+            #endif
+
+            #if !defined(NO_RSA) && defined(USE_CERT_BUFFERS_4096)
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/port/Renesas/renesas_tsip_util.c]: Removed USE_CERT_BUFFER_XXX macro guards.

File: wolfcrypt/src/port/Renesas/renesas_tsip_util.c [hunk 6]
```diff
@@ -3336,6 +3494,379 @@ int wc_tsip_tls_RootCertVerify(
...
+
+#if defined(HAVE_ECC)
+/*   perform signing with the client's ECC private key on hash value of messages
+ *   exchanged with server. 
+ *
+ * parameters
+ *   info->pk.eccsign.in    : the buffer holding hash value of messages
+ *   info->pk.eccsign.inlen : hash data size
+ *   info->pk.eccsign.out   : the buffer where the signature data is output to
+ *   info->pk.eccsign.outlen: the buffer where the length of signature is out
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/port/Renesas/renesas_tsip_util.c]: Reworded the description.

File: wolfcrypt/src/port/Renesas/renesas_tsip_util.c [hunk 7]
```diff
@@ -3336,6 +3494,379 @@ int wc_tsip_tls_RootCertVerify(
...
+            WOLFSSL_MSG("mutex locking error");
+        }
+    }
+    WOLFSSL_LEAVE("tsip_VerifyRsaPkcsCb", ret);
+    return ret;
+}
+#endif /* !NO_RSA */
+
+#if defined(HAVE_ECC)
+/*   perform signing with the client's ECC private key on hash value of messages
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/port/Renesas/renesas_tsip_util.c]: Capitalized it.

File: wolfcrypt/src/port/Renesas/renesas_tsip_util.c [hunk 8]
```diff
@@ -3336,6 +3494,379 @@ int wc_tsip_tls_RootCertVerify(
...
+                        ret = 0;
+                    }
+                    else {
+                        ret = WC_HW_E;
+                    }
+
+                    break;
+                #endif
+
+                #if defined(USE_CERT_BUFFERS_4096)
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/port/Renesas/renesas_tsip_util.c]: Removed USE_CERT_BUFFER_XXX macro guards.

File: wolfcrypt/src/port/Renesas/renesas_tsip_util.c [hunk 9]
```diff
@@ -3336,6 +3494,379 @@ int wc_tsip_tls_RootCertVerify(
...
 #endif /* WOLFSSL_RENESAS_TSIP_TLS */
+#if !defined(NO_RSA)
+/*  perform signing with the client's RSA private key on hash value of messages
+ *  exchanged with server.
+ *
+ * parameters
+ *   info->pk.rsa.in    : not used
+ *   info->pk.rsa.inlen : not used
+ *   info->pk.rsa.out   : the buffer where the signature data is output to
+ *   info->pk.rsa.outlen: the buffer where the length of signature is out
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/port/Renesas/renesas_tsip_util.c]: Reworded the description.

File: wolfcrypt/src/port/Renesas/renesas_tsip_util.c [hunk 10]
```diff
@@ -3336,6 +3494,379 @@ int wc_tsip_tls_RootCertVerify(
     return ret;
 }
 #endif /* WOLFSSL_RENESAS_TSIP_TLS */
+#if !defined(NO_RSA)
+/*  perform signing with the client's RSA private key on hash value of messages
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/port/Renesas/renesas_tsip_util.c]: Capitalized it.

File: wolfcrypt/src/port/Renesas/renesas_tsip_util.c [hunk 11]
```diff
@@ -3336,6 +3494,379 @@ int wc_tsip_tls_RootCertVerify(
...
+                    }
+                    else if (err == TSIP_SUCCESS) {
+                        ret = 0;
+                    }
+                    else {
+                        ret = WC_HW_E;
+                    }
+                    break;
+
+                #if defined(USE_CERT_BUFFERS_1024)
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/port/Renesas/renesas_tsip_util.c]: Removed USE_CERT_BUFFER_XXX macro guards.

**Review by cconlon — APPROVED** [review]

**Pull request merged**